### DISCUSSION TITLE
AuthP Multitenant User Management with Blazor 8

### DISCUSSION THREAD

**Discussion by meliora-solution** [JonPSmith/AuthPermissions.AspNetCore#105] [Ideas]
This project aims to establish a unified user identity for multitenant environments across multiple projects. The implementation relies on AuthPermissions.AspNetCore.

* `TenantCreate.razor` can only be accessed by users who have the `TenantCreate permission`. Therefore, only two users, namely `Super@g1.com` and `AppAdmin@g1.com`, have the authority to create tenants.
* `InviteUser.razor` and `AcceptInvitation.razor` in the `TenantAdminController.` are user interfaces used when the tenant wants users to sign up only by invitation.
* Still working on figuring out the process when a tenant wants users to be able to sign up by themselves.

While not many functions have been created yet, these steps are crucial for implementing multitenancy in IdentityUser.

Project link: `https://github.com/meliora-solution/MultiTenantIdentityUser`

**Comment by meliora-solution** [comment]
Thank you @JonPSmith . The library is very useful.

**Add Tenant in TenantController**

![image](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/5e3dd0a3-427e-47f0-a9c0-13686d9697a5)

There are some good progress implementing this library for Blazor. But I come across a problem with `AuthTenantAdminService`.  

`private readonly IAuthPServiceFactory<ITenantChangeService> _tenantChangeServiceFactory; ` needs to be registered.  In the example3 this registration is done in program.cs

`.RegisterTenantChangeService<InvoiceTenantChangeService>()`

Problem is I dont use Invoice example, so I can not use this way. For temporary solution, I copy AuthTenantAdminService to local clas and removing all the code with `ITenantChangeService`

```
    private readonly IAuthPServiceFactory<ITenantChangeService> _tenantChangeServiceFactory;
   var errorString = await tenantChangeService.CreateNewTenantAsync(newTenantStatus.Result);
   if (errorString != null)
      return status.AddErrorString(this.AlreadyLocalized(), errorString); //we assume the tenantChangeService localized its messages

```

It works, I can create Tenant (Login As Super@g1.com).  But I know this is not a good solution.  Is there any other way to register `ITenantChangeService`.

**Comment by meliora-solution** [comment]
I solved it. I change Example3.InvoiceDbContext.  I remove Invoices and LineInvoice and keep only Companies.  I think I need the company table for detail information about the tenant.

**Comment by meliora-solution** [comment]
I have done all the functions in Example 3.  And Now I am trying to understand example 6.

What I don't understand is about sharding hasowndb=true but the database has 2 tenants.  the owner of db is big rock. But I can assign another tenant to database two.  Why is that ? I thought that if the database has already owned by a tenant then I can not assign to other users ?  
 
![image](https://github.com/JonPSmith/AuthPermissions.AspNetCore/assets/150670966/8b7f8f96-8f6e-49b5-88cc-2bbf633bae09)

**Reply by JonPSmith** [reply]
Hi @meliora-solution,

There two types of sharding: Hybrid or Sharding-Only - see the [Sharding explained](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Sharding-explained) document. Both types have there pros and cons as explained in the document.

- Example 6 shows the Hybrid type 
- Example 7 shows the Sharding-Only type
